src/editorview.cpp

```diff
@@ -422,3 +425,46 @@ void EditorView::setActiveVertex(GripItem* vertex)
     }
   }
 }
+
+GripItem* EditorView::snapGrip() const
+{
+  return m_snapGrip;
+}
+
+EdgeItem* EditorView::snapEdge() const
+{
+  return m_snapEdge;
+}
+
+QPointF EditorView::snapPosition() const
+{
+  return m_snapPosition;
+}
+
+void EditorView::findSnapPosition(const QPoint& pos)
+{
+  m_snapEdge = nullptr;
+  m_snapGrip = nullptr;
+  QPointF mouse = mapToScene(pos);
+  QList<EdgeItem*> edges = itemsInRing<EdgeItem>();
+  qreal closest = std::numeric_limits<qreal>::max();
+  for (EdgeItem* edge : edges) {
+    QPointF snap = edge->nearestPointOnLine(mouse);
+    qreal distance = QLineF(mouse, snap).length();
+    if (distance < closest) {
+      m_snapPosition = snap;
+      m_snapEdge = edge;
+      closest = distance;
+    }
+  }
+  closest = std::numeric_limits<qreal>::max();
+  QList<GripItem*> grips = itemsInRing<GripItem>();
+  for (GripItem* grip : grips) {
```

Comment:
Does this loop actually do anything? How could there be a grip within the ring that wouldn't have been found by being the endpoint of an edge?

Reply:
They are mutually exclusive events. It doesn't use the edge to find grips.

The logic is essentially
```cpp
  if (closestEdge) {
    if (!closestGrip) {
      closestEdge->split(editor->snapPosition());
    }
    else {
      editor->setActiveVertex(closestGrip);
    }
    if (oldActive) {
      mesh->splitPolygon(oldActive, editor->activeVertex());
    }
  }
```
If there is no grip in the ring, insert a vertex in the snapEdge at snapPosition which sets that as the active vertex, If there is a grip in the ring set that as active.

Reply:
Side note: I would like to eventually have it draw a line to the snap marker if it is a valid position to split the polygon but that is outside the scope of this PR

Reply:
*re-reads code a couple times* Oh, I see now.

To be honest, instead of merging the loops, maybe it would be a better idea to split this into two functions, since vertex snapping without edge snapping could be useful. At that point I would probably have the edge snapping function return `QPair<EdgeItem*, QPointF>` instead of storing the results inside the object -- which would allow this function to become `const`. (The vertex snapping function can just return GripItem*.)